test/test_native_mha.py

```diff
@@ -0,0 +1,288 @@
+                    weight_pt[0][-1] = torch.zeros_like(weight_pt[0][-1], device=device, dtype=dtype)
+                    weight_npt[0][-1] = torch.zeros_like(weight_npt[0][-1], device=device, dtype=dtype)
+                    if pad_all:
+                        do_pad_all((weight_pt, weight_npt))
+                else:
+                    for nh in range(num_heads):
+                        weight_pt[0][nh][-1] = torch.zeros_like(weight_pt[0][nh][-1], device=device, dtype=dtype)
+                        weight_npt[0][nh][-1] = torch.zeros_like(weight_npt[0][nh][-1], device=device, dtype=dtype)
+
+        if dtype == torch.half:
+            torch.testing.assert_close(ypt, ynpt, atol=1e-3, rtol=1e-3)
+        else:
+            torch.testing.assert_close(ypt, ynpt)
+
+        if need_weights:
+            torch.testing.assert_close(weight_pt, weight_npt)
+        else:
+            self.assertEqual(weight_pt, weight_npt)
+
+    @dtypesIfCUDA(torch.float, torch.half)
+    @dtypes(torch.float)
+    @skipMeta
+    @torch.inference_mode()
+    def test_native_multihead_self_attention(self, device, dtype):
+        for (use_padding, pad_all) in ((False, False), (True, False), (True, True)):
+            for use_nt in (False, True):
+                # Figuring out exactly which elements of the weights are garbage in this
+                # case eludes me, and it's not particularly enlightening to test anyway
+                # because padding doesn't especially affect the intermediate weights.
+                for need_weights in (False, not pad_all):
+                    for average_attn_weights in (False, True):
+                        self._test_multihead_attention_impl(
+                            device,
+                            dtype,
+                            "self",
+                            use_nt=use_nt,
+                            use_padding=use_padding,
+                            pad_all=pad_all,
+                            need_weights=need_weights,
+                            average_attn_weights=average_attn_weights,
+                        )
+
+    @dtypesIfCUDA(torch.float, torch.half)
+    @dtypes(torch.float)
+    @skipMeta
+    @torch.inference_mode()
+    def test_native_multihead_encoder_decoder_attention(self, device, dtype):
+        self._test_multihead_attention_impl(
+            device,
+            dtype,
+            "encdec",
+            use_nt=False,
+            need_weights=False,
+            average_attn_weights=False,
+        )
+
+    @dtypesIfCUDA(torch.float, torch.half)
+    @dtypes(torch.float)
+    @skipMeta
+    @torch.inference_mode()
+    def test_native_multihead_attention(self, device, dtype):
+        self._test_multihead_attention_impl(
+            device,
+            dtype,
+            "generic",
+            use_nt=False,
+            need_weights=False,
+            average_attn_weights=False,
+        )
+
+
+instantiate_device_type_tests(TestMHADeviceType, globals())
```

Comment:
Are we sure this is running in CI without adding "run_tests" at the end of the file? For example https://github.com/pytorch/pytorch/blob/master/test/test_torch.py#L8387-L8388

Reply:
It does actually. Just that the github view cuts the file before the end.

Reply:
fixed